Python yaml module in ESMX build

There has been an issue in a few places where the ESMX build grabs a different python than one that's already in use and that python doesn't have pyyaml installed so it fails. 

One example of this was in our testing on Orion where our python/yaml based test system was working, but then the ESMX nuopc protos failed because they grabbed a different Python that didn't have yaml. 

Also, one of our collaborators on the SWORD project was having this issue (see email below). (@uturuncoglu  and @anntsay are also interested in this issue from that project.) 

Obviously, one solution is just to install yaml in the python that CMake is grabbing. However, I wonder if it would be possible to have CMake just use the default/current python if one exists, so this doesn't have to be done. 

Thanks!

===== Email =====

Hi Ufuk,

So I installed pyyaml with

pip install pyyaml

and if I try to run the script interactively

python /Users/gtoth/ESMF_nagfor/include/ESMX/Driver/esmx_app_config.py -h

it works fine. If I change the

import yaml

line in  /Users/gtoth/ESMF_nagfor/include/ESMX/Driver/esmx_tools.py to

import yamlxx

then it fails

 File "/Users/gtoth/ESMF_nagfor/include/ESMX/Driver/esmx_tools.py", line 4, in <module>
   import yamlxx
ModuleNotFoundError: No module named 'yamlxx'

so clearly, I have yaml in my Python distribution. Still, if I execute

make test_esmf

it fails with

/Users/gtoth/ESMF_nagfor/bin/bing/Darwin.nag.64.mpich.default/ESMX_Builder --prefix=/Users/gtoth/Tmp_nag/SWMF --verbose
Settings:

 ESMX_BUILD_FILE=esmxBuild.yaml
 ESMF_ESMXDIR=/Users/gtoth/ESMF_nagfor/include/ESMX
 ESMFMKFILE=/Users/gtoth/ESMF_nagfor/esmf.mk
 BUILD_DIR=/Users/gtoth/Tmp_nag/SWMF/ESMF/ESMF_SWMF/src/build
 INSTALL_PREFIX=/Users/gtoth/Tmp_nag/SWMF
 BUILD_TYPE=release
 BUILD_ARGS=
 DISABLE_COMPS=
 BUILD_JOBS=
 MODULEFILE=
 BASHENV=
 TEST=false
 TEST_ARGS=
 VERBOSE=true

-- Detected ESMF Beta snapshot: 0
-- Detected ESMF Beta snapshot: 0
Traceback (most recent call last):
 File "/Users/gtoth/ESMF_nagfor/include/ESMX/Driver/esmx_app_config.py", line 4, in <module>
   from esmx_tools import *
 File "/Users/gtoth/ESMF_nagfor/include/ESMX/Driver/esmx_tools.py", line 4, in <module>
   import yaml
ModuleNotFoundError: No module named 'yaml'
CMake Error at Driver/CMakeLists.txt:12 (message):
 esmx_app_config.py failed processing esmxBuild.yaml
Call Stack (most recent call first):
 Driver/CMakeLists.txt:34 (esmx_check_ret)

so in this case "yaml" is not found. Since I do the make test_esmf as myself, I don't understand how it is succeeding to find yaml in on case and failing in the other. Are you potentially using a fixed python version? For example, if I switch from 3.11 to 3.10 I get

<mesh:~/Tmp_nag/SWMF>python3.10 /Users/gtoth/ESMF_nagfor/include/ESMX/Driver/esmx_app_config.py -h
Traceback (most recent call last):
 File "/Users/gtoth/ESMF_nagfor/include/ESMX/Driver/esmx_app_config.py", line 4, in <module>
   from esmx_tools import *
 File "/Users/gtoth/ESMF_nagfor/include/ESMX/Driver/esmx_tools.py", line 4, in <module>
   import yaml
ModuleNotFoundError: No module named 'yaml'

Clearly, it would be a good idea to use the default python version.

Gabor

--- a/src/addon/ESMX/Driver/CMakeLists.txt
+++ b/src/addon/ESMX/Driver/CMakeLists.txt
@@ -2,6 +2,9 @@ cmake_minimum_required(VERSION 3.22)
 
 # load packages and modules
 list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake)
+set(Python_FIND_UNVERSIONED_NAMES "FIRST")
+set(Python_FIND_FRAMEWORK "LAST")
+set(Python_FIND_REGISTRY  "LAST")
 find_package(Python 3.5 REQUIRED COMPONENTS Interpreter)
 find_package(ESMF 8.6.0 MODULE REQUIRED)
 include(FetchContent)

--- a/src/addon/ESMX/README.md
+++ b/src/addon/ESMX/README.md
@@ -41,6 +41,7 @@ options:
   [--disable-comps=DISABLE_COMPS]
   [--build-type=BUILD_TYPE] or [-g]
   [--build-args=BUILD_ARGS]
+  [--cmake-args=CMAKE_ARGS]
   [--build-jobs=JOBS]
   [--load-modulefile=MODULEFILE]
   [--load-bashenv=BASHENV]
@@ -68,7 +69,10 @@ where:
   --build-type=BUILD_TYPE  build type; valid options are 'debug', 'release', 'relWithDebInfo'
   -g                       (default: release) (-g sets BUILD_TYPE to debug)
 
-  --build-args=BUILD_ARGS  global cmake arguments (e.g. -DVAR=val)
+  --build-args=BUILD_ARGS  build arguments passed to every component (e.g. -DVAR=val)
+
+  --cmake-args=CMAKE_ARGS  cmake arguments are used to configure cmake while building ESMX
+                           (e.g. -DVAR=val)
 
   --build-jobs=BUILD_JOBS  number of jobs used for building esmx and components
 

--- a/src/apps/ESMX_Builder/ESMX_Builder.sh
+++ b/src/apps/ESMX_Builder/ESMX_Builder.sh
@@ -29,7 +29,14 @@ usage () {
   printf "  -g\n"
   printf "      set --build-type=debug\n"
   printf "  --build-args=BUILD_ARGS\n"
-  printf "      cmake arguments (e.g. -DARG=VAL)\n"
+  printf "      build args are passed to each component\n"
+  printf "      arguments are not used to configure ESMX (see --cmake-args)\n"
+  printf "      (e.g. --build-args=\"-DFOO=BAR\")\n"
+  printf "      note: build_args in ESMX_BUILD_FILE supersedes --build-args\n"
+  printf "  --cmake-args=CMAKE_ARGS\n"
+  printf "      cmake args are used to configure cmake while building ESMX\n"
+  printf "      arguments are not passed to components (see --build-args)\n"
+  printf "      (e.g. --cmake-args=\"-DFOO=BAR\")\n"
   printf "  --disable-comps=DISABLE_COMPS\n"
   printf "      disable components\n"
   printf "  --build-jobs=BUILD_JOBS\n"
@@ -63,6 +70,12 @@ settings () {
   printf "  INSTALL_PREFIX=${INSTALL_PREFIX}\n"
   printf "  BUILD_TYPE=${BUILD_TYPE}\n"
   printf "  BUILD_ARGS=${BUILD_ARGS}\n"
+  printf "  CMAKE_ARGS=\n"
+  if [ ${#CMAKE_ARGS[@]} -gt 0 ]; then
+    for arg in "${CMAKE_ARGS[@]}"; do
+      printf "    ${arg}\n"
+    done
+  fi
   printf "  DISABLE_COMPS=${DISABLE_COMPS}\n"
   printf "  BUILD_JOBS=${BUILD_JOBS}\n"
   printf "  MODULEFILE=${MODULEFILE}\n"
@@ -80,6 +93,7 @@ BUILD_FILE="esmxBuild.yaml"
 ESMF_ESMXDIR=""
 BUILD_TYPE="release"
 BUILD_ARGS=""
+CMAKE_ARGS=()
 DISABLE_COMPS=""
 BUILD_JOBS=""
 BUILD_DIR="${CWD}/build"
@@ -125,6 +139,9 @@ while [[ $# -gt 0 ]]; do
     --build-jobs=?*) BUILD_JOBS=${1#*=} ;;
     --build-jobs)  usage_error "$1" "requires an argument" ;;
     --build-jobs=) usage_error "$1" "requires an argument" ;;
+    --cmake-args=?*) CMAKE_ARGS+=(${1#*=}) ;;
+    --cmake-args)  usage_error "$1" "requires an argument" ;;
+    --cmake-args=) usage_error "$1" "requires an argument" ;;
     --disable-comps=?*) DISABLE_COMPS=${1#*=} 
       DISABLE_COMPS=${DISABLE_COMPS/' '/','} 
       DISABLE_COMPS=${DISABLE_COMPS/';'/','} ;;
@@ -231,6 +248,11 @@ if [ ! -z "${BUILD_JOBS}" ]; then
 fi
 if [ ! -z "${BUILD_ARGS}" ]; then
   CMAKE_SETTINGS+=("-DESMX_BUILD_ARGS=${BUILD_ARGS}")
+fi
+if [ ${#CMAKE_ARGS[@]} -gt 0 ]; then
+  for arg in "${CMAKE_ARGS[@]}"; do
+    CMAKE_SETTINGS+=("${arg}")
+  done
 fi
 
 # make settings

